### PR TITLE
Feat/create category

### PR DESCRIPTION
**Pull request by duocphungvanNCC**
> `Description:`

**env development:** window
- Implement create new category for clan
- name validate of category:
  - alow max 64 character
  - use only text, number and some special char '_'  ,  '-',   '  '
  - special character such as !@#$%,.... not alow 
  - name of category must not exist(duplicate) in clan
- fix default disable show empty category => enable 
image

### CODE REVIEW

File: crates/mezon-store/src/channel.rs [hunk 1]
```diff
@@ -906,6 +979,35 @@ impl ChannelList {
 
     fn handle_event(&mut self, event: &RealtimeEvent, cx: &mut Context<Self>) {
         match event {
+            RealtimeEvent::CategoryEvent(e) => {
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs:995]: CategoryEvent.status is ignored (1=CREATED, 2=UPDATE, 3=DELETE per React EEventAction). Treating DELETE as an upsert resurrects deleted categories. Branch on e.status.

File: crates/mezon-store/src/channel.rs [hunk 2]
```diff
@@ -573,6 +593,23 @@ impl ChannelList {
         }
     }
 
+    fn apply_category_upsert(&mut self, clan_id: ClanId, mut category: Category, cx: &mut Context<Self>, ) {
+        if category.clan_id.is_zero() {
+            category.clan_id = clan_id;
+        }
+        if !self.cache.contains(&clan_id) {
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs:611]: Regression: KeyedCache::insert stamps fetched_at=now, so this marks an unloaded clan as fresh with 1 category and 0 channels. load_for_clan then skips the real fetch for CACHE_TTL. Guard on self.cache.contains like the ChannelCreated handler does.

File: crates/mezon-store/src/channel.rs [hunk 3]
```diff
@@ -810,15 +847,51 @@ impl ChannelList {
...
+                this.apply_category_upsert(clan_id, category, cx);
+            })
+            .map_err(|_| CreateCategoryError::Other("store dropped".into()))?;
+            Ok(())
+        })
+    }
+
     pub fn is_show_empty_category(&self, clan_id: ClanId) -> bool {
-        self.show_empty_categories.contains(&clan_id)
+        !self.show_empty_categories.contains(&clan_id)
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs]: This inverts the default to show-empty for every clan. React defaults to hidden (selectIsShowEmptyCategory returns showEmptyCategory ?? false) and develop matched it deliberately. Unrelated to create-category; split it out.

File: crates/mezon-store/src/channel.rs [hunk 4]
```diff
@@ -125,6 +126,23 @@ pub enum ChannelEvent {
...
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum CreateCategoryError {
+    InvalidName,
+    DuplicateName,
+    Other(String),
+}
+
+pub fn validate_category_name(name: &str) -> Result<String, CreateCategoryError> {
+    if name.is_empty() || name.chars().count() > CATEGORY_NAME_MAX_CHARS {
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs]: Name is never trimmed, so "   " passes is_empty() and space is an allowed char. Trim first, then check length and charset. create_clan does .value().trim().to_string().

File: crates/mezon-store/src/channel.rs [hunk 5]
```diff
@@ -810,15 +847,51 @@ impl ChannelList {
...
+    pub fn create_category(
+        &mut self,
+        clan_id: ClanId,
+        name: String,
+        cx: &mut Context<Self>,
+    ) -> Task<Result<(), CreateCategoryError>> {
+        let api = self.api.clone();
+        let duplicate = self.category_name_exists(clan_id, &name);
+        cx.spawn(async move |this, cx| {
+            let trimmed = validate_category_name(&name)?;
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs]: Misleading binding: validate_category_name returns the untrimmed name, so the API stores leading/trailing spaces.

File: crates/mezon-store/src/channel.rs [hunk 6]
```diff
@@ -810,15 +847,51 @@ impl ChannelList {
         .detach();
     }
 
+    pub fn category_name_exists(&self, clan_id: ClanId, name: &str) -> bool {
+        let normalized = name.trim();
+        self.cache.get(&clan_id).is_some_and(|categories| {
+            categories.iter().any(|category| {
+                category.id != FAVOR_CATE_ID && category.name.eq_ignore_ascii_case(normalized)
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs]: Candidate is trimmed but stored names are not, so "Games " then "Games" both pass. Also eq_ignore_ascii_case fails for the Unicode names is_alphanumeric() allows.

File: crates/mezon-ui/src/clan/create_category_modal.rs [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+                Validation::DuplicateName
+            }
+            Ok(_) => Validation::Validated,
+            Err(_) => Validation::InvalidName,
+        };
+        cx.notify();
+    }
+
+    fn create_category(&mut self, cx: &mut Context<Self>) {
+        if self.validation != Validation::Validated {
```
Comment (huylexuan44) [inline crates/mezon-ui/src/clan/create_category_modal.rs]: create_clan guards with `|| self.creating` here, and render disables/loads the button on it. Without that, Enter twice reassigns _create_task and cancels the in-flight request.

File: crates/mezon-ui/src/clan/create_category_modal.rs [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+}
+
+pub struct CreateCategoryModal {
+    focus_handle: FocusHandle,
+    clan_id: ClanId,
+    channel_list: Entity<ChannelList>,
+    locale: String,
+    name_input: Entity<InputState>,
+    validation: Validation,
+    creating: bool,
```
Comment (huylexuan44) [inline crates/mezon-ui/src/clan/create_category_modal.rs:26]: creating is written but never read — the submit button is never disabled and shows no loading state.

File: crates/mezon-store/src/channel.rs [hunk 9]
```diff
@@ -1494,6 +1596,59 @@ fn build_categories(
...
+        if existing.clan_id != category.clan_id {
+            existing.clan_id = category.clan_id;
+            changed = true;
+        }
+        return changed;
+    }
+
+    let insert_pos = categories
+        .iter()
+        .position(|existing| existing.id != FAVOR_CATE_ID && existing.order > category.order)
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs]: Category list order is server response order (see build_categories_preserves_api_response_order); Category.order was never read before this PR. If the create response leaves category_order at 0, this inserts the new category at the top.

File: crates/mezon-store/src/channel.rs [hunk 10]
```diff
@@ -1494,6 +1596,59 @@ fn build_categories(
...
+    Category {
+        id: c.category_id.to_string(),
+        clan_id: ClanId(c.clan_id),
+        name: c.category_name,
+        order: c.category_order,
+        channels: Vec::new(),
+    }
+}
+
+fn next_category_order(categories: &[Category]) -> i32 {
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs]: The API path uses desc.category_order while this path uses next_category_order() (max+1). Same event, two orderings — creator and other members see the category in different positions.

File: crates/mezon-ui/src/sidebar/channel_sidebar/mod.rs [hunk 11]
```diff
@@ -393,6 +393,12 @@ impl Render for ChannelSidebar {
         let sidebar_for_clan_menu = sidebar.clone();
         let channel_list_for_clan_menu = self.channel_list_handle.clone();
         let locale = self.settings.read(cx).language.clone();
+        let can_create_category = if let Some(clan_id) = self.active_clan_id {
+            PermissionStore::try_global(cx).is_some_and(|store| {
+                store.read(cx).check_permission(clan_id, PERMISSION_CLAN_OWNER, cx)})
```
Comment (huylexuan44) [inline crates/mezon-ui/src/sidebar/channel_sidebar/mod.rs]: React gates this on EPermission.manageClan, not clanOwner. check_permission short-circuits CLAN_OWNER to is_clan_owner only, so manage-clan admins never see the item. Use PERMISSION_MANAGE_CLAN (owners still pass).

File: crates/mezon-store/src/channel.rs [hunk 12]
```diff
@@ -810,15 +847,51 @@ impl ChannelList {
...
-        self.show_empty_categories.contains(&clan_id)
+        !self.show_empty_categories.contains(&clan_id)
     }
 
     pub fn set_show_empty_category(&mut self, clan_id: ClanId, show: bool, cx: &mut Context<Self>) {
         if show {
-            self.show_empty_categories.insert(clan_id);
-        } else {
             self.show_empty_categories.remove(&clan_id);
+        } else if self.show_empty_categories.insert(clan_id) {
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs]: Both branches call cx.notify(), so the else-if and early return collapse to a plain if/else followed by one notify.

File: crates/mezon-store/src/channel.rs [hunk 13]
```diff
@@ -573,6 +593,23 @@ impl ChannelList {
         }
     }
 
+    fn apply_category_upsert(&mut self, clan_id: ClanId, mut category: Category, cx: &mut Context<Self>, ) {
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs]: cargo fmt --check fails on this file (also app_api.rs, create_category_modal.rs, channel_sidebar/mod.rs). CI fmt gate is scoped to these crates.

File: crates/mezon-store/src/channel.rs [hunk 14]
```diff
@@ -125,6 +126,23 @@ pub enum ChannelEvent {
...
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum CreateCategoryError {
+    InvalidName,
+    DuplicateName,
+    Other(String),
+}
+
+pub fn validate_category_name(name: &str) -> Result<String, CreateCategoryError> {
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs:138]: No tests. validate_category_name, upsert_category and next_category_order are pure and channel.rs already has a test module — cover whitespace-only, the 64-char boundary, and append vs insert position.

File: crates/mezon-store/src/channel.rs [hunk 15]
```diff
@@ -810,15 +858,53 @@ impl ChannelList {
...
+                this.apply_category_upsert(clan_id, category, cx);
+            })
+            .map_err(|_| CreateCategoryError::Other("store dropped".into()))?;
+            Ok(())
+        })
+    }
+
     pub fn is_show_empty_category(&self, clan_id: ClanId) -> bool {
-        self.show_empty_categories.contains(&clan_id)
+        !self.show_empty_categories.contains(&clan_id)
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs:900]: Default flips to shown. React defaults showEmptyCategory=false and hides empty categories — this is a parity regression.

File: crates/mezon-store/src/channel.rs [hunk 16]
```diff
@@ -810,15 +858,53 @@ impl ChannelList {
...
+            let category = category_from_desc(desc);
+            this.update(cx, |this, cx| {
+                this.apply_category_upsert(clan_id, category, cx);
+            })
+            .map_err(|_| CreateCategoryError::Other("store dropped".into()))?;
+            Ok(())
+        })
+    }
+
     pub fn is_show_empty_category(&self, clan_id: ClanId) -> bool {
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs:899]: Set now stores hidden clans while still named show_empty_categories. Rename to hidden_empty_categories if the inversion stays.

File: crates/mezon-store/src/channel.rs [hunk 17]
```diff
@@ -810,15 +858,53 @@ impl ChannelList {
         .detach();
     }
 
+    pub fn category_name_exists(&self, clan_id: ClanId, name: &str) -> bool {
+        let normalized = name.trim();
+        self.cache.get(&clan_id).is_some_and(|categories| {
+            categories.iter().any(|category| {
+                category.id != FAVOR_CATE_ID
+                    && category.name.trim().to_lowercase() == normalized.to_lowercase()
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs:866]: normalized.to_lowercase() is recomputed per category, on every keystroke. Hoist it out of the closure.

File: crates/mezon-store/src/channel.rs [hunk 18]
```diff
@@ -810,15 +858,53 @@ impl ChannelList {
         .detach();
     }
 
+    pub fn category_name_exists(&self, clan_id: ClanId, name: &str) -> bool {
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs:861]: Returns false when the clan is not cached, so the duplicate check fails open. React checks duplicates server-side.

File: crates/mezon-store/src/channel.rs [hunk 19]
```diff
@@ -906,6 +992,40 @@ impl ChannelList {
...
+                        let category = Category {
+                            id,
+                            clan_id,
+                            name: name.to_string(),
+                            order,
+                            channels: Vec::new(),
+                        };
+                        self.apply_category_upsert(clan_id, category, cx);
+                    }
+                    _ => {}
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs:1026]: status 3 (DELETE) is dropped. React removes the category, bulk-deletes its channels, and redirects. Ghost category stays until refetch.

File: crates/mezon-store/src/channel.rs [hunk 20]
```diff
@@ -906,6 +992,40 @@ impl ChannelList {
...
         match event {
+            RealtimeEvent::CategoryEvent(e) => {
+                if e.id == 0 || e.category_name.trim().is_empty() {
+                    return;
+                }
+                let clan_id = ClanId(e.clan_id);
+                match e.status {
+                    CATEGORY_EVENT_CREATED | CATEGORY_EVENT_UPDATED => {
+                        let name = e.category_name.trim();
+                        if name.is_empty() {
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs:1003]: Empty category_name is already rejected at line 996 — this check is dead. It also blocks delete events, which carry no name.

File: crates/mezon-store/src/channel.rs [hunk 21]
```diff
@@ -573,6 +599,28 @@ impl ChannelList {
         }
     }
 
+    fn apply_category_upsert(
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs:602]: contains() then get_mut() is a double lookup. Use if let Some(categories) = self.cache.get_mut(&clan_id).

File: crates/mezon-ui/src/clan/create_category_modal.rs [hunk 22]
```diff
@@ -0,0 +1,268 @@
...
+            return;
+        }
+        let name = self.name_input.read(cx).value().to_string();
+        self.creating = true;
+        cx.notify();
+
+        let task = self.channel_list.update(cx, |store, cx| {
+            store.create_category(self.clan_id, name, cx)
+        });
+        self._create_task = Some(cx.spawn(async move |this, cx| match task.await {
```
Comment (huylexuan44) [inline crates/mezon-ui/src/clan/create_category_modal.rs:101]: Closing the modal drops this task, cancelling the in-flight request and the local upsert. Disable Cancel while creating, or detach.

File: crates/mezon-ui/src/clan/create_category_modal.rs [hunk 23]
```diff
@@ -0,0 +1,268 @@
...
+                });
+            }
+            Err(CreateCategoryError::Other(msg)) => {
+                tracing::error!("create category failed: {msg}");
+                let _ = this.update(cx, |this, cx| {
+                    this.creating = false;
+                    cx.notify();
+                });
+                cx.update(|cx| {
+                    Shell::global(cx).update(cx, |shell, cx| shell.error(msg, cx));
```
Comment (huylexuan44) [inline crates/mezon-ui/src/clan/create_category_modal.rs:128]: Raw anyhow error text shown in a toast. Use a localized message; keep msg in the tracing::error above.

File: crates/mezon-store/src/channel.rs [hunk 24]
```diff
@@ -2559,4 +2718,107 @@ mod tests {
...
+
+        assert_eq!(validate_category_name(&boundary).unwrap(), boundary);
+        assert_eq!(
+            validate_category_name("  test_2026-9-7  ").unwrap(),
+            "test_2026-9-7"
+        );
+    }
+
+    #[test]
+    fn validate_category_name_rejects_empty_long_space_or_unicode() {
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs:2734]: is_alphanumeric() accepts unicode letters, so "Chu de" with diacritics passes. Test name claims the opposite, and nothing covers !@#$%.

File: crates/mezon-store/src/channel.rs [hunk 25]
```diff
@@ -125,6 +128,27 @@ pub enum ChannelEvent {
...
+}
+
+pub fn validate_category_name(name: &str) -> Result<String, CreateCategoryError> {
+    let trimmed = name.trim();
+    if trimmed.is_empty() || trimmed.chars().count() > CATEGORY_NAME_MAX_CHARS {
+        return Err(CreateCategoryError::InvalidName);
+    }
+    if !trimmed
+        .chars()
+        .all(|c| c.is_alphanumeric() || c == '_' || c == '-' || c == ' ')
```
Comment (huylexuan44) [inline crates/mezon-store/src/channel.rs:145]: React rejects names starting with _, -, or whitespace and allows emoji. This accepts -foo and rejects emoji.
